ztoc/ztoc.go

```diff
@@ -253,8 +253,23 @@ func (zt Ztoc) ExtractFromTarGz(gz string, filename string) (string, error) {
 	return string(bytes), nil
 }
 
-// Zinfo deserilizes and returns a Zinfo based on the zinfo bytes and compression
-// algorithm in the ztoc.
-func (zt Ztoc) Zinfo() (compression.Zinfo, error) {
-	return compression.NewZinfo(zt.CompressionAlgorithm, zt.Checkpoints)
+func (zt Ztoc) ConsumeTOC() {
+	zt.TOC.FileMetadata = nil
+}
+
+// ConsumeZinfo deserilizes and returns a Zinfo based on the zinfo bytes and compression
+// algorithm in the ztoc. This process nils the byte array used in the process, so
+// it can only be called once.
+func (zt Ztoc) ConsumeZinfo() (compression.Zinfo, error) {
+	if zt.Checkpoints != nil {
+		zinfo, err := compression.NewZinfo(zt.CompressionAlgorithm, zt.Checkpoints)
+		if err != nil {
+			return nil, err
+		}
+		// Zinfo holds all of the checkpoint data, so we can free the raw Checkpoints array.
+		zt.Checkpoints = nil
```

Comment:
Does this acutally work? The methods exposed by Ztoc take non-pointer receivers, so we are modifying the deferenced _copy_  of the Ztoc. 

eg: https://go.dev/play/p/Ng5JA8n3m06

I think while we work on stabilizing the Ztoc API, it should be up to consumers to manually free the checkpoint bytes. 

```go

type SpanManager struct {
	zinfo                             compression.Zinfo
	ztoc                              *ztoc.Ztoc
}

func New(ztoc *ztoc.Ztoc, r *io.SectionReader, cache cache.BlobCache, retries int, cacheOpt ...cache.Option) *SpanManager {
       index, err := ztoc.Zinfo()
	if err != nil {
		return nil
	}
        ztoc.Checkpoints = nil
	m := &SpanManager{
		zinfo:                             index,
		ztoc:                              ztoc,
                ...
	}
}
```

Reply:
> Does this acutally work? The methods exposed by Ztoc take non-pointer receivers, so we are modifying the deferenced copy of the Ztoc.

Ah, good catch. Would it be considered poor practice then to have some methods refer to pointers, while others do not? 

> it should be up to consumers to manually free the checkpoint bytes

Should it, though? The checkpoint bytes are essentially garbage for anything outside of creating the Zinfo, so I don't think it's a bad thing to have the Zinfo auto-consume. Though, if it's bad practice to refer to some methods by reference and others by value, then I don't think we have a choice outside of what you mentioned.

Reply:
Yeah, mixed receivers are generally considered bad practice. I suppose we could change all Ztoc methods to take pointer receivers, but that's _technically_ considered an API change and so we _may_ want to think about what a stable access API looks like before that, although I'm not strongly convinced by that either.

Reply:
> I suppose we could change all Ztoc methods to take pointer receivers

I was going to pitch changing everything to be a pointer, but then functions such as ExtractFile would consume the zinfo, and doesn't pass the zinfo back, which really doesn't make sense.

I think I'll just revert a bunch of these changes and set the Checkpoints/TOC to nil manually. But I want to stress our need to come back to this soon.